keps/sig-apps/4368-support-managed-by-label-for-batch-jobs/README.md

```diff
@@ -0,0 +1,848 @@
+
+- passing custom parameters to the external controller
+
+## Proposal
+
+The proposal is to support the `managed-by` label to disabled the Job controller
+from synchronizing jobs with the annotation.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a developer of Kueue I want to have Job API which allows me to implement the
+MultiKueue design. For this reason I need a way to disable the main Job
+controller on the management cluster.
+
+The mechanism should be per-Job, because the management cluster might also be
+one of the worker clusters, for two reasons:
+1. Disabling the Job controller per cluster requires access the `kube-controller-manager`
+   manifest. Such access is generally discouraged by cloud providers.
+2. The management cluster may also be a worker. Supporting this scenario is important
+   for smooth transition of Kueue users from a single-cluster to multi-cluster.
+
+Ideally, I prefer the mechanism to be lightweight so that it is easy to be adopted
+by other Job CRDs supported by Kueue
+(see [here](https://github.com/kubernetes-sigs/kueue/blob/6d428f3279a9ca0e204c083dc649dbbc6558db71/config/components/manager/controller_manager_config.yaml#L31-L42)):
+MPIJob, RayJob, JobSet, multiple Kubeflow jobs.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Prior work
+
+This approach of allowing another controller to mirror information between APIs
+is already supported with the `managed-by` label used by
+EndpointSlices (`endpointslice.kubernetes.io/managed-by`) and IPAddresses
+(`ipaddress.kubernetes.io/managed-by`).
+
+#### Graduating directly to Beta
+
+The base-functionality implementation is very simple, just short-circuit the
+`syncJob` invocation. Also, since the feature is just label-based, there is
+technically no need to go via Alpha in order to ensure that all
+kube-apiserver instances (in HA setup, see [here](https://kubernetes.io/releases/version-skew-policy/#supported-version-skew))
+recognize the new API, as in case of new fields. A similar approach was used
+to graduate the
+[Pod Index Label for StatefulSets and Indexed Jobs](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/4017-pod-index-label)
+feature with similar characteristics.
+
+Finally, we would like to have the feature soak without unnecessary delay.
+
+#### Using label selectors
+
+We consider using label selectors by the Job controller to identify the subset
+of jobs it should watch. However, due to backwards-compatibility there are two
+cases:
+1. old jobs without the label
+2. new jobs with the label equal to `job-controller`
+
+Using label selectors might be relevant in case of [Story 1](#story-1), as
+MultiKueue can have many jobs in the management cluster. Keeping the jobs in
+the Job controller's informer, and starting the `syncJob` function for them is
+wasteful.
+
+However, the label selector semantics does not support OR-ing these two cases.
+For this reason, we need to ensure all jobs have the label set. One mechanism
```

Comment:
Why? Can't we just exclude those jobs that have managed-by labels set to some different value?

Reply:
We can select jobs to be excluded (not synced) by the label selector `managed-by!=job-controller`. But I don't think there is a syntax to "negate it", because we care about the jobs we want to sync ( no `managed-by` OR `managed-by=job-controller`).

Anyway, we give up on the idea of using label selectors, because the ROI is low. I'm discussing this in alternatives.